src/common/src/mlib/str.h

```diff
@@ -565,4 +642,488 @@ mstr_contains_any_of(mstr_view str, mstr_view needle)
+ * and a size `len`. If not null, the pointer `data` points to an array of mutable
+ * `char` of length `len + 1`, where the character at `data[len]` is always zero,
+ * and must not be modified.
+ *
+ * @note The string MAY contain nul (zero-value) characters, so using them with
+ * C string APIs could truncate unexpectedly.
+ * @note The string itself may be "null" if the `data` member of the string is
+ * a null pointer. A zero-initialized `mstr` is null. The null string is distinct
+ * from the empty string, which has a non-null `.data` that points to an empty
+ * C string.
+ */
+typedef struct mstr {
+   /**
+    * @brief Pointer to the first char in the string, or NULL if
+    * the string is null.
+    *
+    * The pointed-to character array has a length of `len + 1`, where
+    * the character at `data[len]` is always null.
+    *
+    * @warning Attempting to overwrite the null character at `data[len]`
+    * will result in undefined behavior!
+    *
+    * @note An empty string is not equivalent to a null string! An empty string
+    * will still point to an array of length 1, where the only char is the null
+    * terminator.
+    */
+   char *data;
+   /**
+    * @brief The number of characters in the array pointed-to by `data`
+    * that precede the null terminator.
+    */
+   size_t len;
+} mstr;
+
+
+/**
+ * @brief Resize an existing or null `mstr`, without initializing any of the
+ * added content other than the null terminator. This operation is potientially
+ * UNSAFE, because it gives uninitialized memory to the caller.
+ *
+ * @param str Pointer to a valid `mstr`, or a null `mstr`.
+ * @param new_len The new length of the string.
+ * @return true If the operation succeeds
+ * @return false Otherwise
+ *
+ * If `str` is a null string, this function will initialize a new `mstr` object
+ * on-the-fly.
+ *
+ * If the operation increases the length of the string (or initializes a new string),
+ * then the new `char` in `str.data[str.len : new_len] will contain uninitialized
+ * values. The char at `str.data[new_len]` WILL be set to zero, to ensure there
+ * is a null terminator. The caller should always initialize the new string
+ * content to ensure that the string has a specified value.
+ */
+static inline bool
+mstr_resize_for_overwrite(mstr *const str, const size_t new_len)
+{
+   // We need to allocate one additional char to hold the null terminator
+   size_t alloc_size = new_len;
+   if (mlib_add(&alloc_size, 1) || alloc_size > PTRDIFF_MAX) {
+      // Allocation size is too large
+      return false;
+   }
+   // Try to (re)allocate the region
+   char *data = (char *)realloc(str->data, alloc_size);
+   if (!data) {
+      // Failed to (re)allocate
+      return false;
+   }
+   // Note: We do not initialize any of the data in the newly allocated region.
+   // We only set the null terminator. It is up to the caller to do the rest of
+   // the init.
+   data[new_len] = 0;
```

Comment:
```suggestion
   data[new_len] = '\0';
```
`char` literal for `char` assignment.